Stock.playground/Contents.swift

```diff
@@ -3,116 +3,147 @@ import Foundation
-        return shareValue * sharesAccrued
+        let percentageAccrued = Double(months) / Double(endMonths)
+        let sharesAccrued = floor(percentageAccrued * Double(shares))
+        return valueAt(price, sharesAccrued: sharesAccrued)
     }
 
-    func monthsSince(date: GrantDate) -> Int
+    func valueAt(_ price: Double, date: Date) -> Double
     {
-        let futureDate = date.foundationDate
-        let grantDate = self.grantDate.foundationDate
-
-        let dateComponents = NSCalendar.currentCalendar().components(NSCalendarUnit.Month, fromDate: grantDate, toDate: futureDate, options: NSCalendarOptions(rawValue: 0))
-        return dateComponents.month
+        let monthsElapsed = date.monthsSince(grantDate)
+        if (monthsElapsed < 0) { return 0 }
+        return valueAt(price, monthsElapsed: UInt(monthsElapsed))
     }
 
-    func valueBy(date: GrantDate, atPrice: Double) -> Double
+    private func valueAt(_ price: Double, sharesAccrued: Double) -> Double
     {
-        return self.valueAfter(self.monthsSince(date), atPrice: atPrice)
+        let shareValue = price - strikePrice
+        let value = max(0, shareValue)
+        let shareCount = max(0, min(Double(shares), sharesAccrued))
+        return value * shareCount
     }
+}
+
 
-    func valuesBy(dates: [GrantDate], atPrice: Double) -> [Double]
+extension Date
+{
+    static func grantDate(month: Int, day: Int, year: Int) -> Date
     {
-        return dates.map({ self.valueBy($0, atPrice: atPrice) })
+        let signing = NSDateComponents.init()
+        signing.calendar = Calendar.grantDateCalendar()
+        signing.month = month
+        signing.day = day
+        signing.year = year
+        return signing.date!
     }
 
-    func curriedValueBy(date: GrantDate) -> (Double) -> Double
+    func monthsSince(_ date: Date) -> Int
     {
-        return { price in return self.valueAfter(self.monthsSince(date), atPrice: price) }
+        let dateComponents = Calendar.grantDateCalendar().components(.month, from: date, to: self)
+        return dateComponents.month!
     }
 
-    func curriedValueAt(price: Double) -> (GrantDate) -> Double
+    func after(months: Int) -> Date
     {
-        return { date in return self.valueAfter(self.monthsSince(date), atPrice: price) }
+        return after(months, unit: .month)
     }
-}
-
 
-let signing = StockGrant.GrantDate(month: 1, day: 1, year: 2016)
-let signingGrant = StockGrant(shares: 20000, grantDate: signing, cliffMonths: 12, endMonths: 48, strikePrice: 1.22)
+    func after(years: Int) -> Date
+    {
+        return after(years, unit: .year)
+    }
 
-// Two years at $22 (using months or StockGrant.GrantDate)
-let twoYears = StockGrant.GrantDate(month: 1, day: 1, year: 2018)
-signingGrant.valueBy(twoYears, atPrice: 22)
-signingGrant.valueAfter(24, atPrice: 22)
+    private func after(_ amount: Int, unit: Calendar.Unit) -> Date
```

Comment:
Might be best to make a tuple `typealias TimeMeasurement = (count: Int, calendarUnit: Calendar.Unit)` and use that here. Then the method could be `after(_ amount: TimeMeasurement)` or some such.